### PR TITLE
use produce-consume model for stixshifter transmission and translation

### PR DESCRIPTION
**Pull request by tingdai**

### CODE REVIEW

File: src/kestrel/datasource/manager.py
Comment (subbyte) [inline src/kestrel/datasource/manager.py]: How about we make it generic and test any connector with [inspect.iscoroutinefunction](https://docs.python.org/3/library/inspect.html#inspect.iscoroutinefunction)?

File: src/kestrel/datasource/manager.py
Comment (subbyte) [inline src/kestrel/datasource/manager.py:8]: May move the `nest_asyncio` here as the first time `asyncio.run()` is called:
- https://github.com/opencybersecurityalliance/kestrel-lang/issues/306
- https://github.com/opencybersecurityalliance/kestrel-lang/blob/develop/src/kestrel_datasource_stixshifter/interface.py#L113

File: src/kestrel_datasource_stixshifter/interface.py
Comment (subbyte) [inline src/kestrel_datasource_stixshifter/interface.py]: Any idea why we do the single cycle loop? May consider put the number of consumers as a parameter in the configuration?

File: src/kestrel_datasource_stixshifter/interface.py
Comment (subbyte) [inline src/kestrel_datasource_stixshifter/interface.py]: Shall we just create a function `make_ingest_stixbundle_filepath(batch_index)` instead of manipulating the string through the module? If someone forget to replace substring in the future when using `ingestfile`, that may cause issue.

**Review by subbyte — COMMENTED** [review]
Thanks @tingdai for the first step to put Kestrel into async mode! The producer-consumer model looks good. I have some suggestions on improvements before merge.

File: src/kestrel_datasource_stixshifter/config.py
Comment (subbyte) [inline src/kestrel_datasource_stixshifter/config.py]: Suggest to put the config:
```
options:
    async_translation_workers_count: 1
```
Basically rename `consumes` to `workers` since it may be producers in the future for the second queue, the preffix `async` describes the scope, which is not multi-processing.

We may also update `TRANSLATION_CONSUME_NUM` to `ASYNC_TRANSLATION_WORKERS_CNT`.

File: src/kestrel/utils.py
Comment (subbyte) [inline src/kestrel/utils.py]: May remove the function if not used.

File: src/kestrel/utils.py
Comment (subbyte) [inline src/kestrel/utils.py]: Could we move this function from `/src/kestrel/*` to `/src/kestrel_datasource_stixshifter/*`? These are two Python packages, which should have minimal dependencies. You may just put it in `interface.py` if the function is only used once. Or a `utils.py` module if the function is used by multiple modules in `kestrel_datasource_stixshifter`.

File: src/kestrel_datasource_stixshifter/interface.py
Comment (subbyte) [inline src/kestrel_datasource_stixshifter/interface.py]: Could we combine the two lines into the `make_ingest_stixbundle_filepath()` function:
```
ingestfile = ingestdir / f"{i}_batch_index_{data_path_striped}.json"
ingest_stixbundle_filepath = make_ingest_stixbundle_filepath(ingestfile)
```
and not use `replace()`?

File: src/kestrel_datasource_stixshifter/interface.py
Comment (subbyte) [inline src/kestrel_datasource_stixshifter/interface.py:285]: Do we still have `connector_results` prepared for `fast_translation()`?

**Review by subbyte — CHANGES_REQUESTED** [review]

File: src/kestrel_datasource_stixshifter/config.py
Comment (subbyte) [inline src/kestrel_datasource_stixshifter/config.py]: Why do we add the `kestrel_datasource_stixshifter` package to the stix-shifter debug logging settings?

File: src/kestrel/datasource/manager.py
Comment (subbyte) [inline src/kestrel/datasource/manager.py:42]: May need to do `run_until_complete()`, need a second opinion on the logic.

File: src/kestrel_datasource_stixshifter/interface.py
Comment (subbyte) [inline src/kestrel_datasource_stixshifter/interface.py:393]: Need to check with @pcoccoli whether we have an async version of `cache()` in `firepit.aio`.

**Review by subbyte — COMMENTED** [review]
Excellent flow now!

I have some minor comments to discuss today, plus a generic one: should we rename all `translate_consume` and similar functions to `translate_n_ingest_consume`?

File: src/kestrel_datasource_stixshifter/interface.py
Comment (pcoccoli) [reply src/kestrel_datasource_stixshifter/interface.py:393]: Yes, this will work.

File: src/kestrel_datasource_stixshifter/interface.py
Comment (pcoccoli) [inline src/kestrel_datasource_stixshifter/interface.py]: Please remove the print now

File: src/kestrel/datasource/manager.py
Comment (subbyte) [reply src/kestrel/datasource/manager.py:42]: Confirmed this is no longer an issue for Python>=3.7 where Kestrel currently supports. The `run()` function calls `run_until_complete()` internally: https://github.com/python/cpython/blob/3.7/Lib/asyncio/runners.py#L43

File: src/kestrel_datasource_stixshifter/interface.py
Comment (subbyte) [inline src/kestrel_datasource_stixshifter/interface.py]: Need to be removed.

File: src/kestrel_datasource_stixshifter/config.py
Comment (subbyte) [reply src/kestrel_datasource_stixshifter/config.py]: Actually you find a bug (not related to what you are doing) in Kestrel CLI---when `--debug` flag is used, there is no issue, but when the environment variable `KESTREL_DEBUG` is used, the debug level is not set by the [CLI](https://github.com/opencybersecurityalliance/kestrel-lang/blob/develop/src/kestrel/__main__.py).

In conclusion, it is safe to remove this line. I will fix the bug #334 separately. If you test with Kestrel CLI and `--debug` flag, it should behave as expected.

**Review by subbyte — APPROVED** [review]

**Review by pcoccoli — APPROVED** [review]

File: src/kestrel_datasource_stixshifter/interface.py
Comment (tingdai) [reply src/kestrel_datasource_stixshifter/interface.py]: It is removed in the latest commit https://github.com/opencybersecurityalliance/kestrel-lang/pull/316/commits/817298e23031ab6a1d9cae72b5b6fc7b5eb6becf

**Pull request merged**
